lib/rules/prefer-destructuring.js

```diff
@@ -0,0 +1,128 @@
+        // Helpers
+        //--------------------------------------------------------------------------
+
+        /**
+         * Determines if the given node node is accessing an array index
+         *
+         * This is used to differentiate array index access from object property
+         * access.
+         *
+         * @param {ASTNode} node the node to evaluate
+         * @returns {boolean} whether or not the node is an integer
+         */
+        function isArrayIndexAccess(node) {
+            return Number.isInteger(node.property.value);
+        }
+
+        /**
+         * Check if a given variable declarator is coming from an property access
+         * that should be using destructuring instead
+         *
+         * @param {ASTNode} node the variable declarator to check
+         * @returns {void}
+         */
+        function checkVariableDeclarator(node) {
+
+            // Skip if variable is declared without assignment
+            if (!node.init) {
+                return;
+            }
+
+            if (node.init.type === "Identifier") {
+                const objectExpression = node.id.type === "ObjectPattern" && node.id;
+
+                if (objectExpression) {
+                    const prop = objectExpression.properties[0];
+                    const key = prop.key;
+                    const value = prop.value;
+
+                    if ((key.name === value.name) && (key.start !== value.start)) {
+                        context.report({ node: prop, message: "Unnecessary duplicate variable name" });
+                    }
+                }
+            }
+
+            // We only care about member expressions past this point
+            if (node.init.type !== "MemberExpression") {
+                return;
+            }
+
+            const memberExpression = node.init;
+
+            if (checkArrays && isArrayIndexAccess(memberExpression)) {
+                context.report({ node, message: "Use array destructuring" });
+                return;
+            }
+
+            if (checkObjects && !isArrayIndexAccess(memberExpression)) {
+                const variableIdentifier = node.id;
+                const property = node.init.property;
+
+                if (property.type === "Literal" && variableIdentifier.name === property.value) {
+                    context.report({ node, message: "Use object destructuring" });
+                }
+
+                if (property.type === "Identifier" && variableIdentifier.name === property.name) {
+                    context.report({ node, message: "Use object destructuring" });
+                }
+            }
+        }
+
+        //--------------------------------------------------------------------------
+        // Public
+        //--------------------------------------------------------------------------
+
+        return {
+            VariableDeclarator: checkVariableDeclarator
```

Comment:
It looks like this only checks `VariableDeclarator` nodes, but I think it should also check `AssignmentExpression` nodes.

For example, the following should get reported as an error:

```js
foo = bar[0];
```

Since it could get replaced with:

```js
[foo] = bar;
```

This should also get reported:

```js
foo = bar.baz;
```

because it could be replaced with:

```js
({baz: foo} = bar);
```